lib/msal-core/src/Configuration.ts

```diff
@@ -0,0 +1,181 @@
+ * storeAuthStateInCookie   - If set, the library will store the auth request state required for validation of the auth flows in the browser cookies.
+ *                              By default this flag is set to false.
+ */
+export type CacheOptions = {
+  cacheLocation?: CacheLocation;
+  storeAuthStateInCookie?: boolean;
+};
+
+/**
+ * Library Specific Options
+ *
+ * logger                       - Used to initialize the Logger object; TODO: Expand on logger details or link to the documentation on logger
+ * loadFrameTimeout             - maximum time the library should wait for a frame to load
+ * tokenRenewalOffsetSeconds    - sets the window of offset needed to renew the token before expiry
+ *
+ */
+export type SystemOptions = {
+  logger?: Logger;
+  loadFrameTimeout?: number;
+  tokenRenewalOffsetSeconds?: number;
+};
+
+/**
+ * App/Framework specific environment Support
+ *
+ * isAngular                - flag set to determine if it is Angular Framework. Used to broadcast tokens. TODO: detangle this dependency from core.
+ * unprotectedResources     - Array of URI's which are unprotected resources. MSAL will not attach a token to outgoing requests that have these URI. Defaults to 'null'.
+ * protectedResourceMap     - This is mapping of resources to scopes used by MSAL for automatically attaching access tokens in web API calls.
+ *                              A single access token is obtained for the resource. So you can map a specific resource path as follows:
+ *                              {"https://graph.microsoft.com/v1.0/me", ["user.read"]},
+ *                              or the app URL of the resource as: {"https://graph.microsoft.com/", ["user.read", "mail.send"]}.
+ *                              This is required for CORS calls.
+ *
+ */
+export type FrameworkOptions = {
+  isAngular?: boolean;
+  unprotectedResources?: Array<string>;
+  protectedResourceMap?: Map<string, Array<string>>;
+};
+
+/**
+ * Configuration Object
+ */
+export type Configuration = {
+  auth: AuthOptions,
+  cache?: CacheOptions,
+  system?: SystemOptions,
+  framework?: FrameworkOptions
+};
+
+/**
+ * Function to set the default options when not explicitly set
+ *
+ * @param TAuthOptions
+ * @param TCacheOptions
+ * @param TSystemOptions
+ * @param TFrameworkOptions
+ *
+ * @returns TConfiguration object
+ */
+
+// destructure with default settings
+export function buildConfiguration(
+  {
+    clientId = "",
+    authority = null,
+    validateAuthority = true,
+    redirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    postLogoutRedirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    state = "",
+    navigateToLoginRequestUrl = true
+  }: AuthOptions = { clientId },
+  {
+    cacheLocation = "sessionStorage",
+    storeAuthStateInCookie = false
+  }: CacheOptions = {},
+  {
+    logger = new Logger(null),
```

Comment:
is this in the case where a user might pass in their own logger? what if they pass in a logger that doesn't have the same api as ours ? 
What if they pass in a logger that does have the `logPii` function and expose pii information ?

Reply:
They have to adhere to our api today. It is a mutually agreed upon structure as it stands. This PR is just mimicking what is working, an overhaul in future is already in the pipeline from what I understand.

Reply:
When they do not pass a logger, we initialize it for them with our defaults.

Reply:
Resolving this conversation. Please tag this with the upcoming logging testing.